### ISSUE TITLE
loss decreased too slow,spend 2 days to train efficient-d4，is it normal？

### ISSUE THREAD

**Issue by Aries5522** [xuannianz/EfficientDet#42]

**Comment by ammaradel** [comment]
Same thing here on a custom dataset. Did it give good results eventually after 2 days?

**Comment by ammaradel** [comment]
This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the  mAP never exceeded 0.03.

I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).

**Comment by Aries5522** [comment]
> This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> 
> I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).

thank you  i will try  efficient-D5 these days

**Comment by Aries5522** [comment]
> This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> 
> I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).

Do u add bifpn？？Does it work？

**Comment by ammaradel** [comment]
> > This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> > I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).
> 
> Do u add bifpn？？Does it work？

No, I didn't try that option yet.

**Comment by jyc8251** [comment]
> This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> 
> I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).

Can you share the setting to run efficient-D5 ? For example which optimizer you are using, batch size, learning rate, etc.
Thanks!

**Comment by ammaradel** [comment]
> > This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> > I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).
> 
> Can you share the setting to run efficient-D5 ? For example which optimizer you are using, batch size, learning rate, etc.
> Thanks!

I used the default optimizer (Adam) and learning rate (1e-3). I don't remember the batch size but I think it was 2 or 4.

The problem now is that, when I tested it on another dataset, I got the previous problem (the loss almost doesn't decrease). I tested several learning rates with freeze-backbone and freeze-bn.

I wonder if the problem is linked to the batch size. I don't have enough memory to train it with a batch size > 4.

**Comment by jyc8251** [comment]
> > > This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> > > I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).
> > 
> > 
> > Can you share the setting to run efficient-D5 ? For example which optimizer you are using, batch size, learning rate, etc.
> > Thanks!
> 
> I used the default optimizer (Adam) and learning rate (1e-3). I don't remember the batch size but I think it was 2 or 4.
> 
> The problem now is that, when I tested it on another dataset, I got the previous problem (the loss almost doesn't decrease). I tested several learning rates with freeze-backbone and freeze-bn.
> 
> I wonder if the problem is linked to the batch size. I don't have enough memory to train it with a batch size > 4.

That problem is mostly linked to the batch size. The "loss almost doesn't decrease' issue was solved when I put in more gpus (thus more ram, and bigger batch size) to train. From my own experience the batch size can not be no less than 16.

**Comment by vincentz0602** [comment]
> > > > This happened to me with efficient-D4. The loss decreased very little even after 500 epochs. And the mAP never exceeded 0.03.
> > > > I'm running efficient-D5 now and the loss is decreasing relatively fast. The mAP on the same dataset attained 0.14 after only 4 epochs of step 1 (with --freeze-backbone).
> > > 
> > > 
> > > Can you share the setting to run efficient-D5 ? For example which optimizer you are using, batch size, learning rate, etc.
> > > Thanks!
> > 
> > 
> > I used the default optimizer (Adam) and learning rate (1e-3). I don't remember the batch size but I think it was 2 or 4.
> > The problem now is that, when I tested it on another dataset, I got the previous problem (the loss almost doesn't decrease). I tested several learning rates with freeze-backbone and freeze-bn.
> > I wonder if the problem is linked to the batch size. I don't have enough memory to train it with a batch size > 4.
> 
> That problem is mostly linked to the batch size. The "loss almost doesn't decrease' issue was solved when I put in more gpus (thus more ram, and bigger batch size) to train. From my own experience the batch size can not be no less than 16.

have you ever tried smaller learning rate? 
This happened to me when I trained d0, initial learning rate was 1e-3, and then gradient vanishing happened, result was push to the saturated zone of the sigmoid.
